src/client/Microsoft.Identity.Client/ManagedIdentity/AppServiceManagedIdentitySource.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    internal class AppServiceManagedIdentitySource : ManagedIdentitySource
+    {
+        // MSI Constants. Docs for MSI are available here https://docs.microsoft.com/azure/app-service/overview-managed-identity
+        protected const string AppServiceMsiApiVersion = "2019-08-01";
+        protected const string SecretHeaderName = "X-IDENTITY-HEADER";
+        protected const string ClientIdHeaderName = "client_id";
+
+        private const string MsiEndpointInvalidUriError = "The environment variable IDENTITY_ENDPOINT contains an invalid Uri.";
+
+        private readonly Uri _endpoint;
+        private readonly string _secret;
+        private readonly string _clientId;
+        private readonly string _resourceId;
+
+        public static ManagedIdentitySource TryCreate(RequestContext requestContext)
+        {
+            var msiSecret = EnvironmentVariables.IdentityHeader;
+            return TryValidateEnvVars(EnvironmentVariables.IdentityEndpoint, msiSecret, out Uri endpointUri)
+                ? new AppServiceManagedIdentitySource(requestContext, endpointUri, msiSecret)
+                : null;
+        }
+
+        protected static bool TryValidateEnvVars(string msiEndpoint, string secret, out Uri endpointUri)
+        {
+            endpointUri = null;
+
+            // if BOTH the env vars endpoint and secret values are null, this MSI provider is unavailable.
+            if (string.IsNullOrEmpty(msiEndpoint) || string.IsNullOrEmpty(secret))
+            {
+                return false;
+            }
+
+            try
+            {
+                endpointUri = new Uri(msiEndpoint);
+            }
+            catch (FormatException ex)
+            {
+                throw new MsalClientException(MsalError.InvalidManagedIdentityEndpoint, MsiEndpointInvalidUriError, ex);
```

Comment:
TryX methods shouldn't throw; should return false here. Or make it into normal Validate method and throw if validation fails.
A more general question - if we fail to create a managed identity source for whatever reason, do we log and return null or throw and catch in the MIClient? (I'm leaning towards returning null)

Reply:
Normally I agree, but I think the code is copied from https://github.com/Azure/azure-sdk-for-net/blob/7ce7c69eb805731fc057d8ae1f1948a6903622f1/sdk/identity/Azure.Identity/src/AppServiceManagedIdentitySource.cs#L23 ... so in case the msiEndpoint is badly formatted, they bail.

Reply:
I guess, is your intention to keep the same public end-to-end behavior for MSI functionality so AzureSDK has minimal work to implement this?

Reply:
It is important to throw here to indicate that this resource has endpoints available but was not able to create relevant endpoint. If we return false here then as we add more managed identity sources, it will try to create the next managed identity source. We don't want to try create any other resource here.